lerobot/scripts/server/policy_server.py

```diff
@@ -0,0 +1,429 @@
+            TimedAction(t_0 + i * environment_dt, action, i_0 + i) for i, action in enumerate(action_chunk)
+        ]
+
+    @torch.no_grad()
+    def _run_act_policy(self, observation: dict[str, torch.Tensor]) -> torch.Tensor:
+        """Run ACT-like policies"""
+        start_time = time.time()
+
+        # prepare observation for policy forward pass
+        batch = self.policy.normalize_inputs(observation)
+        normalize_time = time.time()
+        self.logger.debug(f"Observation normalization time: {normalize_time - start_time:.6f}s")
+
+        if self.policy.config.image_features:
+            batch = dict(batch)  # shallow copy so that adding a key doesn't modify the original
+            batch["observation.images"] = [batch[key] for key in self.policy.config.image_features]
+            prep_time = time.time()
+            self.logger.debug(f"Observation image preparation time: {prep_time - normalize_time:.6f}s")
+
+        # forward pass outputs up to policy.config.n_action_steps != actions_per_chunk
+        actions = self.policy.model(batch)[0][:, : self.actions_per_chunk]
+
+        actions = self.policy.unnormalize_outputs({"action": actions})["action"]
+
+        end_time = time.time()
+        self.logger.info(f"[ACT] Action chunk generation total time: {end_time - start_time:.6f}s")
+
+        return actions
+
+    @torch.no_grad()
+    def _run_pi0_policy(self, observation: dict[str, torch.Tensor]) -> torch.Tensor:
+        """Run PI0-like policies"""
+        raise NotImplementedError("PI0 policy not implemented yet")
+
+    @torch.no_grad()
+    def _run_smolvla_policy(
+        self, observation: dict[str, torch.Tensor], noise: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        """Run smolvla-like policies"""
+        observation = self.policy.normalize_inputs(observation)
+
+        images, img_masks = self.policy.prepare_images(observation)
+        state = self.policy.prepare_state(observation)
+        lang_tokens, lang_masks = self.policy.prepare_language(observation)
+
+        actions = self.policy.model.sample_actions(
+            images, img_masks, lang_tokens, lang_masks, state, noise=noise
+        )
+
+        # Unpad actions
+        original_action_dim = self.policy.config.action_feature.shape[0]
+        actions = actions[:, :, :original_action_dim]
+
+        actions = self.policy.unnormalize_outputs(
+            {"action": actions, "robot_type": [self.policy.config.robot_type]}
+        )["action"]
+
+        return actions
+
+    def _get_action_chunk(
+        self, observation: dict[str, torch.Tensor], policy_type: str = "act"
+    ) -> torch.Tensor:
+        """Get an action chunk from the policy"""
+        if policy_type == "act":
+            return self._run_act_policy(observation)
+        elif policy_type == "smolvla":
+            return self._run_smolvla_policy(observation)
+        else:
+            raise ValueError(f"Policy class {policy_type} not supported")
+
```

Comment:
1. Naive question: why not use the `select_action` method? If I understand correctly, you have almost the same code, but you don't add to the action queue.

2. We need zero coupling, the server shouldn't know about policy internals. We can find a way to decouple this. Let's talk about it in front of the whiteboard :)